test/subscription_map.hpp

```diff
@@ -38,28 +38,97 @@
  *           Using a published topic, we can find all topic filters which match the specified topic
  *       . A stored topic map, storing topic -> data
  *           Using a new topic filter, we can find all stored topics which match the specified topic filter
+ *
+ *   Subscription map stores all entries in a tree
+ *   the tree starts from a root node, and topic filters are tokenized and stored in the tree
+ *
+ *   For example if the topic_filter example/monitor/Clients is stored, the following nodes are created:
+ *   root -> example -> monitor -> Clients
+ *
+ *   Every node in the tree may store one or multiple subscribers. Nodes store a reference count to the number of subscribers
+ *   so for example, if we store the following topic_filters:
+ *     example/
+ *     example/monitor/Clients
+ *
+ *   the subscription map looks as follows:
+ *     root(2) -> example(2) -> monitor(1) -> Clients (1)
+ *
+ *   hash and + are stored as normal nodes within the tree, but the parent node knows if a hash child is available. This
+ *   improves the matching, no extra lookup is required to see if a # or + child is available in a child node:
+ *
+ *     example/#
+ *
+ *    stores the following tree:
+ *      root -> example (hash: yes) -> #
+ *
+ *    and
+ *
+ *      example/+
+ *
+ *    stores the following tree:
+ *      root -> example (plus: yes) -> #
+ *
+ *    all node entries are stored in a single hash map. The key for every node is: (parent node id, path)
+ *
+ *      so if we store: root/example/test
+ *      root (id:1) -> example (id:2, key:1,example) -> test (id:3, key:2,test)
+ *
+ *    also, every node stores the key of its parent, allowing quick traversing from leaf to root of the tree
  */
 
+// Combined storage for count and flags
+// we can have 32bit or 64bit version
+template<std::size_t N>
+struct count_storage
```

Comment:
I recommend making the default specialization have no members, so it causes a compiler error.

Then specialize explicitly on 4 and 8 so they are the only two valid instantiations.

Reply:
I think so too.
@kleunen , could you update the code?

I think that the declaration only primary template is good. See https://wandbox.org/permlink/4LlquRZs2moxlDLw

Reply:
I would recommend against this and would recommend making 64 bits the default. I would say, mqtt cpp now only compiles on 32bit and 64bit targets. We use a 32bit struct as a performance optimization for 32 bit architectures: reduce memory consumption mostly. If you would compile for another platform: 64 bit or possible future 128bit systems? or 16bit system? I would recommend just using the 64bit identifier as a default (it is now 32 bit). Higher does not make that much sense. 

If you have a declaration only target, you would get a compile error. But the compile error does not say that much. Choosing a default implementation (64bit) and continue compilation would be fine.

Reply:
It also makes sense to only have a 64bit version. Because the maximum number of entries in the subscription map is different now for 32bit and 64bit architerctures. Not sure if this is really what you want. My reasoning behind is: 32bit is probably some embedded architecture and not a server and number of subscriptions is probably limited.

Reply:
I disagree.
At least it is bad N is ignored.
Is `count_storage` always used with explicit template argument ? I guess so.
Compile error is better than unexpected success.

Reply:
> It also makes sense to only have a 64bit version. Because the maximum number of entries in the subscription map is different now for 32bit and 64bit architerctures. Not sure if this is really what you want. My reasoning behind is: 32bit is probably some embedded architecture and not a server and number of subscriptions is probably limited.

@kleunen 

If you are concerned about not having a clear error message, you can do a static-assertion in the default, non-specialized, version of the struct.

Something like

```
template<size_t N>
struct mystruct
{
    static_assert(N == 4 || N == 8, "mystruct only knows how to handle architectures with 32 or 64 bit size_ts. Please update me to support your platform.");
};
```

Reply:
That is a good suggestion. That is my concern